File: lib/models/pet.dart

class Pet {
  final String id;
  final String? name;
  final String breed;
  final String color;
  final String? size;
  final String? age;
  final String? gender;
  final String? description;
  final List<String> imageUrls;
  final List<double>? imageEmbeddings;
  final String? thumbnailUrl;
  final double latitude;
  final double longitude;
  final String? locationAddress;
  final String? locationCity;
  final String? locationState;
  final String? locationCountry;
  final String status; // "lost" or "found"
  final DateTime dateLostOrFound;
  final Map<String, dynamic> contactInfo;
  final String ownerId;
  final double? imageQualityScore;
  final bool isDuplicate;
  final String? duplicateOf;
  final int matchCount;
  final DateTime? lastMatched;
  final String? distinctiveFeatures;
  final String? medicalInfo;
  final String? behaviorTraits;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isActive;

  Pet({
    required this.id,
    this.name,
    required this.breed,
    required this.color,
    this.size,
    this.age,
    this.gender,
    this.description,
    required this.imageUrls,
    this.imageEmbeddings,
    this.thumbnailUrl,
    required this.latitude,
    required this.longitude,
    this.locationAddress,
    this.locationCity,
    this.locationState,
    this.locationCountry,
    required this.status,
    required this.dateLostOrFound,
    required this.contactInfo,
    required this.ownerId,
    this.imageQualityScore,
    this.isDuplicate = false,
    this.duplicateOf,
    this.matchCount = 0,
    this.lastMatched,
    this.distinctiveFeatures,
    this.medicalInfo,
    this.behaviorTraits,
    required this.createdAt,
    required this.updatedAt,
    this.isActive = true,
  });

  factory Pet.fromJson(Map<String, dynamic> json) {
    return Pet(
      id: json['id'] as String,
      name: json['name'] as String?,
      breed: json['breed'] as String,
      color: json['color'] as String,
      size: json['size'] as String?,
      age: json['age'] as String?,
      gender: json['gender'] as String?,
      description: json['description'] as String?,
      imageUrls: List<String>.from(json['image_urls'] as List),
      imageEmbeddings: json['image_embeddings'] != null
          ? List<double>.from(json['image_embeddings'] as List)
          : null,
      thumbnailUrl: json['thumbnail_url'] as String?,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      locationAddress: json['location_address'] as String?,
      locationCity: json['location_city'] as String?,
      locationState: json['location_state'] as String?,
      locationCountry: json['location_country'] as String?,
      status: json['status'] as String,
      dateLostOrFound: DateTime.parse(json['date_lost_or_found'] as String),
      contactInfo: Map<String, dynamic>.from(json['contact_info'] as Map? ?? {}),
      ownerId: json['owner_id'] as String,
      imageQualityScore: json['image_quality_score'] as double?,
      isDuplicate: json['is_duplicate'] as bool? ?? false,
      duplicateOf: json['duplicate_of'] as String?,
      matchCount: json['match_count'] as int? ?? 0,
      lastMatched: json['last_matched'] != null
          ? DateTime.parse(json['last_matched'] as String)
          : null,
      distinctiveFeatures: json['distinctive_features'] as String?,
      medicalInfo: json['medical_info'] as String?,
      behaviorTraits: json['behavior_traits'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'breed': breed,
      'color': color,
      'size': size,
      'age': age,
      'gender': gender,
      'description': description,
      'image_urls': imageUrls,
      'image_embeddings': imageEmbeddings,
      'thumbnail_url': thumbnailUrl,
      'latitude': latitude,
      'longitude': longitude,
      'location_address': locationAddress,
      'location_city': locationCity,
      'location_state': locationState,
      'location_country': locationCountry,
      'status': status,
      'date_lost_or_found': dateLostOrFound.toIso8601String(),
      'contact_info': contactInfo,
      'owner_id': ownerId,
      'image_quality_score': imageQualityScore,
      'is_duplicate': isDuplicate,
      'duplicate_of': duplicateOf,
      'match_count': matchCount,
      'last_matched': lastMatched?.toIso8601String(),
      'distinctive_features': distinctiveFeatures,
      'medical_info': medicalInfo,
      'behavior_traits': behaviorTraits,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'is_active': isActive,
    };
  }

  Pet copyWith({
    String? id,
    String? name,
    String? breed,
    String? color,
    String? size,
    String? age,
    String? gender,
    String? description,
    List<String>? imageUrls,
    List<double>? imageEmbeddings,
    String? thumbnailUrl,
    double? latitude,
    double? longitude,
    String? locationAddress,
    String? locationCity,
    String? locationState,
    String? locationCountry,
    String? status,
    DateTime? dateLostOrFound,
    Map<String, dynamic>? contactInfo,
    String? ownerId,
    double? imageQualityScore,
    bool? isDuplicate,
    String? duplicateOf,
    int? matchCount,
    DateTime? lastMatched,
    String? distinctiveFeatures,
    String? medicalInfo,
    String? behaviorTraits,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isActive,
  }) {
    return Pet(
      id: id ?? this.id,
      name: name ?? this.name,
      breed: breed ?? this.breed,
      color: color ?? this.color,
      size: size ?? this.size,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      description: description ?? this.description,
      imageUrls: imageUrls ?? this.imageUrls,
      imageEmbeddings: imageEmbeddings ?? this.imageEmbeddings,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      locationAddress: locationAddress ?? this.locationAddress,
      locationCity: locationCity ?? this.locationCity,
      locationState: locationState ?? this.locationState,
      locationCountry: locationCountry ?? this.locationCountry,
      status: status ?? this.status,
      dateLostOrFound: dateLostOrFound ?? this.dateLostOrFound,
      contactInfo: contactInfo ?? this.contactInfo,
      ownerId: ownerId ?? this.ownerId,
      imageQualityScore: imageQualityScore ?? this.imageQualityScore,
      isDuplicate: isDuplicate ?? this.isDuplicate,
      duplicateOf: duplicateOf ?? this.duplicateOf,
      matchCount: matchCount ?? this.matchCount,
      lastMatched: lastMatched ?? this.lastMatched,
      distinctiveFeatures: distinctiveFeatures ?? this.distinctiveFeatures,
      medicalInfo: medicalInfo ?? this.medicalInfo,
      behaviorTraits: behaviorTraits ?? this.behaviorTraits,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Pet &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'Pet{id: $id, name: $name, breed: $breed, status: $status}';
  }
}

class PetCreate {
  final String? name;
  final String breed;
  final String color;
  final String? size;
  final String? age;
  final String? gender;
  final String? description;
  final List<String> imageUrls;
  final double latitude;
  final double longitude;
  final String? locationAddress;
  final String? locationCity;
  final String? locationState;
  final String? locationCountry;
  final String status;
  final DateTime? dateLostOrFound;
  final Map<String, dynamic> contactInfo;
  final String? distinctiveFeatures;
  final String? medicalInfo;
  final String? behaviorTraits;

  PetCreate({
    this.name,
    required this.breed,
    required this.color,
    this.size,
    this.age,
    this.gender,
    this.description,
    required this.imageUrls,
    required this.latitude,
    required this.longitude,
    this.locationAddress,
    this.locationCity,
    this.locationState,
    this.locationCountry,
    required this.status,
    this.dateLostOrFound,
    this.contactInfo = const {},
    this.distinctiveFeatures,
    this.medicalInfo,
    this.behaviorTraits,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'breed': breed,
      'color': color,
      'size': size,
      'age': age,
      'gender': gender,
      'description': description,
      'image_urls': imageUrls,
      'latitude': latitude,
      'longitude': longitude,
      'location_address': locationAddress,
      'location_city': locationCity,
      'location_state': locationState,
      'location_country': locationCountry,
      'status': status,
      'date_lost_or_found': dateLostOrFound?.toIso8601String(),
      'contact_info': contactInfo,
      'distinctive_features': distinctiveFeatures,
      'medical_info': medicalInfo,
      'behavior_traits': behaviorTraits,
    };
  }
}

class PetUpdate {
  final String? name;
  final String? breed;
  final String? color;
  final String? size;
  final String? age;
  final String? gender;
  final String? description;
  final List<String>? imageUrls;
  final double? latitude;
  final double? longitude;
  final String? locationAddress;
  final String? locationCity;
  final String? locationState;
  final String? locationCountry;
  final String? status;
  final Map<String, dynamic>? contactInfo;
  final String? distinctiveFeatures;
  final String? medicalInfo;
  final String? behaviorTraits;
  final bool? isActive;

  PetUpdate({
    this.name,
    this.breed,
    this.color,
    this.size,
    this.age,
    this.gender,
    this.description,
    this.imageUrls,
    this.latitude,
    this.longitude,
    this.locationAddress,
    this.locationCity,
    this.locationState,
    this.locationCountry,
    this.status,
    this.contactInfo,
    this.distinctiveFeatures,
    this.medicalInfo,
    this.behaviorTraits,
    this.isActive,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    
    if (name != null) data['name'] = name;
    if (breed != null) data['breed'] = breed;
    if (color != null) data['color'] = color;
    if (size != null) data['size'] = size;
    if (age != null) data['age'] = age;
    if (gender != null) data['gender'] = gender;
    if (description != null) data['description'] = description;
    if (imageUrls != null) data['image_urls'] = imageUrls;
    if (latitude != null) data['latitude'] = latitude;
    if (longitude != null) data['longitude'] = longitude;
    if (locationAddress != null) data['location_address'] = locationAddress;
    if (locationCity != null) data['location_city'] = locationCity;
    if (locationState != null) data['location_state'] = locationState;
    if (locationCountry != null) data['location_country'] = locationCountry;
    if (status != null) data['status'] = status;
    if (contactInfo != null) data['contact_info'] = contactInfo;
    if (distinctiveFeatures != null) data['distinctive_features'] = distinctiveFeatures;
    if (medicalInfo != null) data['medical_info'] = medicalInfo;
    if (behaviorTraits != null) data['behavior_traits'] = behaviorTraits;
    if (isActive != null) data['is_active'] = isActive;
    
    return data;
  }
}

File: backend/models/pet.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from bson import ObjectId
from .user import PyObjectId

class Pet(BaseModel):
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    name: Optional[str] = None
    breed: str
    color: str
    size: Optional[str] = None  # small, medium, large
    age: Optional[str] = None  # puppy, young, adult, senior
    gender: Optional[str] = None  # male, female, unknown
    description: Optional[str] = None
    image_urls: List[str] = []
    image_embeddings: Optional[List[float]] = None
    thumbnail_url: Optional[str] = None
    
    # Location information
    latitude: float
    longitude: float
    location_address: Optional[str] = None
    location_city: Optional[str] = None
    location_state: Optional[str] = None
    location_country: Optional[str] = None
    
    # Status and metadata
    status: str  # "lost" or "found"
    date_lost_or_found: datetime = Field(default_factory=datetime.utcnow)
    contact_info: Dict[str, Any] = {}
    owner_id: PyObjectId
    
    # Quality and duplicate detection
    image_quality_score: Optional[float] = None
    is_duplicate: bool = False
    duplicate_of: Optional[PyObjectId] = None
    
    # Matching information
    match_count: int = 0
    last_matched: Optional[datetime] = None
    
    # Additional metadata
    distinctive_features: Optional[str] = None
    medical_info: Optional[str] = None
    behavior_traits: Optional[str] = None
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    is_active: bool = True

    @validator('status')
    def validate_status(cls, v):
        if v.lower() not in ['lost', 'found']:
            raise ValueError('Status must be either "lost" or "found"')
        return v.lower()

    @validator('image_urls')
    def validate_images(cls, v):
        if not v:
            raise ValueError('At least one image URL is required')
        return v

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}

class PetCreate(BaseModel):
    name: Optional[str] = None
    breed: str
    color: str
    size: Optional[str] = None
    age: Optional[str] = None
    gender: Optional[str] = None
    description: Optional[str] = None
    image_urls: List[str]
    latitude: float
    longitude: float
    location_address: Optional[str] = None
    location_city: Optional[str] = None
    location_state: Optional[str] = None
    location_country: Optional[str] = None
    status: str
    date_lost_or_found: Optional[datetime] = None
    contact_info: Dict[str, Any] = {}
    distinctive_features: Optional[str] = None
    medical_info: Optional[str] = None
    behavior_traits: Optional[str] = None

class PetUpdate(BaseModel):
    name: Optional[str] = None
    breed: Optional[str] = None
    color: Optional[str] = None
    size: Optional[str] = None
    age: Optional[str] = None
    gender: Optional[str] = None
    description: Optional[str] = None
    image_urls: Optional[List[str]] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    location_address: Optional[str] = None
    location_city: Optional[str] = None
    location_state: Optional[str] = None
    location_country: Optional[str] = None
    status: Optional[str] = None
    contact_info: Optional[Dict[str, Any]] = None
    distinctive_features: Optional[str] = None
    medical_info: Optional[str] = None
    behavior_traits: Optional[str] = None
    is_active: Optional[bool] = None

class PetResponse(BaseModel):
    id: str
    name: Optional[str] = None
    breed: str
    color: str
    size: Optional[str] = None
    age: Optional[str] = None
    gender: Optional[str] = None
    description: Optional[str] = None
    image_urls: List[str]
    thumbnail_url: Optional[str] = None
    latitude: float
    longitude: float
    location_address: Optional[str] = None
    location_city: Optional[str] = None
    location_state: Optional[str] = None
    location_country: Optional[str] = None
    status: str
    date_lost_or_found: datetime
    contact_info: Dict[str, Any]
    distinctive_features: Optional[str] = None
    medical_info: Optional[str] = None
    behavior_traits: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    is_active: bool

    class Config:
        from_attributes = True
